### PR TITLE
dev/core#587 Permissions on GroupContact and Relationship API calls

### PR DESCRIPTION
**Pull request by pradpnayak**
Overview
----------------------------------------
https://lab.civicrm.org/dev/core/issues/587

Before
----------------------------------------
Doesn't respect ACL.

After
----------------------------------------
Respect ACL.

### CODE REVIEW

**Comment by pradpnayak** [comment]
Will add unit test soon.

File: api/v3/Relationship.php [hunk 1]
```diff
@@ -74,6 +74,12 @@ function civicrm_api3_relationship_delete($params) {
     return civicrm_api3_create_error('Invalid value for relationship ID');
   }
 
+  // Check if contact can be modified.
+  _civicrm_api3_check_edit_permissions(
```
Comment (eileenmcnaughton) [inline api/v3/Relationship.php]: We shouldn't need this here as basic_create does this

**Comment by eileenmcnaughton** [comment]
@pradpnayak you read my mind :-) - I think the api_v3_ACLPermissionTest class is a good starting point.

I flicked through this change & it all makes sense to me - aside from the one comment I made

**Comment by MegaphoneJon** [comment]
I did an `r-run` on this and it seems to be good!  I'm happy to run this in production for a bit before it gets merged if that would make folks happier.

**Comment by MegaphoneJon** [comment]
Confirming that this has been in place since late May on all my production sites without issues.

**Comment by MegaphoneJon** [comment]
@pradpnayak Style issues

**Comment by mlutfy** [comment]
jenkins, test this please

**Comment by mlutfy** [comment]
@pradpnayak Any luck with writing a test?

Test fail:

```
  error: patch failed: CRM/Contact/BAO/Relationship.php:2339                               
  error: CRM/Contact/BAO/Relationship.php: patch does not apply       
```

**Comment by JoeMurray** [comment]
Monish, could you rebase and get tests to pass then resubmit for QA/Review, please?

Thanks.

**Comment by eileenmcnaughton** [comment]
@totten @colemanw you might want to eyeball this - it's apiv3 but you might want to check there is no interaction with your 'relationship vortex' plan

It has a test & I'm assuming @monishdeb is intending on reviewing & merging in the near future.

**Comment by monishdeb** [comment]
I have added the unit-test to capture the EDIT permission impact on groupContact and relationship add/update action.

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by eileenmcnaughton** [comment]
Actually there is a problem here I think (@colemanw ) - by reducing the permission level in CRM/Core/Permission it is reduced for apiv3 AND v4. However, the implementation only puts permissions BACK for v3 so apiv4 would have a new permission bypass if we merge this 

ping @pfigel @seamuslee001

**Comment by eileenmcnaughton** [comment]
I'm going to close this as I think it will cause a regression for apiv4 as is - we should focus on the apiv4 version at this stage

**Comment by masetto** [comment]
On CiviCRM 5.65.2 I still have this problem:  

image

The group being removed from the contact was created by the administrator and it is not a reserved group.  
The user has "edit groups" permssion but he absolutely cannot have "edit all contacts".  

I don't understand what should be done to solve it.

**Comment by colemanw** [comment]
This was fixed for APIv4 in https://github.com/civicrm/civicrm-core/pull/24186, however the "Remove from group" button on the contact summary tab still uses APIv3.
https://github.com/civicrm/civicrm-core/blob/06a8e7e398293495ee026035214e75125f10dacb/templates/CRM/Contact/Page/View/GroupContact.tpl#L208

If someone wanted to they could attempt to refactor that code to use v4. Personally I think the effort is better spent replacing that whole tab with SearchKit, as I've just done [for the notes tab](https://github.com/civicrm/civicrm-core/pull/27610).

**Pull request closed without merging**